blazor/datagrid/connecting-to-data-source/microsoft-sql-server.md

```diff
@@ -0,0 +1,641 @@
+           {
+               // Filtering
+               DataSource = DataOperations.PerformFiltering(DataSource, dm.Where, dm.Where[0].Operator);
+           }
+           int count = DataSource.Cast<Order>().Count();
+           return dm.RequiresCounts ? new DataResult() { Result = DataSource, Count = count } : (object)DataSource;
+       }
+
+{% endhighlight %}
+
+**Perform Sorting:** 
+When utilizing a custom adaptor, managing the sorting operation involves overriding the `Read` or `ReadAsync` method of the **DataAdaptor** abstract class.
+
+In the code example below, sorting a custom data source can be accomplished by employing the built-in `PerformSorting` method of the `DataOperations` class. Alternatively, you can implement your own method for sorting operations and bind the resulting data to the grid.
+
+{% highlight razor %}
+
+   public class CustomAdaptor : DataAdaptor
+   {
+       public OrderData Order = new OrderData();
+       // Performs data Read operation
+       public override async Task<object> ReadAsync(DataManagerRequest dm, string key = null)
+       {
+           IEnumerable<Order> DataSource = await Order.GetOrdersAsync();    
+           
+              if (dm.Sorted != null && dm.Sorted.Count > 0)
+           {
+               // Sorting
+               DataSource = DataOperations.PerformSorting(DataSource, dm.Sorted);
+           }
+           int count = DataSource.Cast<Order>().Count();
+           return dm.RequiresCounts ? new DataResult() { Result = DataSource, Count = count } : (object)DataSource;
+       }
+
+{% endhighlight %}
+
+**Perform Grouping:** 
+
+When employing a Custom Adaptor, the grouping operation must be managed within the `Read` or `ReadAsync` method of the Custom adaptor.
+
+The provided sample code illustrates how to implement the grouping operation within a Custom Adaptor.
+
+{% highlight razor %}
+
+   public class CustomAdaptor : DataAdaptor
+   {
+       public OrderData Order = new OrderData();
+       // Performs data Read operation
+       public override async Task<object> ReadAsync(DataManagerRequest dm, string key = null)
+       {
+           IEnumerable<Order> DataSource = await Order.GetOrdersAsync();    
+           int count = DataSource.Cast<Order>().Count();
+           DataResult DataObject = new DataResult();
+           if (dm.Group != null)
+           {
+              IEnumerable ResultData = DataSource.ToList();
+               // Grouping
+               foreach (var group in dm.Group)
+               {
+                   ResultData = DataUtil.Group<Order>(ResultData, group, dm.Aggregates, 0, dm.GroupByFormatter);
+               }
+               DataObject.Result = ResultData;
+               DataObject.Count = count;
+               return dm.RequiresCounts ? DataObject : (object)ResultData;
+           }
+
+           return dm.RequiresCounts ? new DataResult() { Result = DataSource, Count = count } : (object)DataSource;
+       }
+{% endhighlight %}
+
```

Comment:
Add notes and mention the order that we need to follow to have a better performance